### PR TITLE
Build and package framework dependent package

### PR DESCRIPTION
**Pull request by adityapatwardhan**
## PR Summary

Add the functionality to build a framework dependent (shared framework) package for PowerShell.

The changes create two packages, one for Windows and other for Linux, due to `#if` code.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: tools/packaging/packaging.psm1 [hunk 1]
```diff
@@ -124,12 +139,21 @@ function Start-PSPackage {
             # also ensure `Start-PSPackage` does what the user asks/expects, because once packages
             # are generated, it'll be hard to verify if they were built from the correct content.
             $params = @('-Clean')
-            $params += '-CrossGen'
+
+            if ($Type -ne 'fxdependent') {
+                $params += '-CrossGen'
```
Comment (TravisEz13) [inline tools/packaging/packaging.psm1]: We should make sure Steve and Joey are aware of this

Comment (adityapatwardhan) [reply tools/packaging/packaging.psm1]: @SteveL-MSFT @joeyaiello CrossGen cannot be done as this is a common package for x86 and x64 and CrossGen is architecture specific.

File: build.psm1 [hunk 2]
```diff
@@ -3067,7 +3081,7 @@ function New-PSOptionsObject
         [String]
         $Top,
 
-        [Parameter(Mandatory)]
+        ## This is non-mandatory for fxdependent package.
```
Comment (TravisEz13) [inline build.psm1]: Is this still true?

**Review by TravisEz13 — APPROVED** [review]
nothing blocking

**Comment by dantraMSFT** [comment]
Although the changes look 'reasonable', it's difficult to determine if the results actually work for the intended purpose.  Is there any way we can create a test verify the result?

**Comment by anmenaga** [comment]
UpdatableHelpSystem failures in PowerShell-CI-linux.

File: tools/packaging/packaging.psm1 (hunk 1 above)
Comment (SteveL-MSFT) [reply tools/packaging/packaging.psm1]: Let's add a comment here.

File: build.psm1 (hunk 2 above)
Comment (adityapatwardhan) [reply build.psm1]: Good catch. Added back Mandatory.

File: tools/releaseBuild/Images/microsoft_powershell_windowsservercore/PowerShellPackage.ps1 [hunk 3]
```diff
@@ -116,6 +132,19 @@ try{
             Copy-Item -Path $file -Destination "$destination\" -Force
         }
     }
+    elseif (!$ComponentRegistration.IsPresent -and $Runtime -eq 'fxdependent')
```
[the remark this answers is not in the document]
Comment (adityapatwardhan) [reply tools/releaseBuild/Images/microsoft_powershell_windowsservercore/PowerShellPackage.ps1]: Fixed.

File: tools/packaging/packaging.psm1 [hunk 4]
```diff
@@ -105,14 +110,24 @@ function Start-PSPackage {
...
 
+        if($Type -eq 'fxdependent') {
+            ## We do not check for runtime and crossgen for framework dependent package.
+            $precheckFailed = -not $Script:Options -or                          ## Start-PSBuild hasn't been executed yet
+                        -not $PSModuleRestoreCorrect -or                        ## Last build didn't specify '-PSModuleRestore' correctly
+                        $Script:Options.Configuration -ne $Configuration -or    ## Last build was with configuration other than 'Release'
+                        $Script:Options.Framework -ne "netcoreapp2.1"           ## Last build wasn't for CoreCLR
+        }
+        else {
+            $precheckFailed = -not $Script:Options -or                                ## Start-PSBuild hasn't been executed yet
```
[the remark this answers is not in the document]
Comment (adityapatwardhan) [reply tools/packaging/packaging.psm1]: Fixed.

File: tools/packaging/packaging.psm1 [hunk 5]
```diff
@@ -124,12 +139,21 @@ function Start-PSPackage {
...
+            if ($Type -ne 'fxdependent') {
+                $params += '-CrossGen'
+            }
+
             if (!$IncludeSymbols.IsPresent) {
                 $params += '-PSModuleRestore'
             }
 
-            $params += '-Runtime', $Runtime
+            if($Type -eq 'fxdependent') {
```
[the remark this answers is not in the document]
Comment (adityapatwardhan) [reply tools/packaging/packaging.psm1]: Fixed.

**Comment by adityapatwardhan** [comment]
@anmenaga  The failures are due to "Network unreachable" issue when downloading from download center. Other help tests work as they are downloading help from Azure blob storage.

**Comment by adityapatwardhan** [comment]
@dantraMSFT I have filed an issue to tracking adding new tests for framework dependent package #7750

**Review by anmenaga — APPROVED** [review]

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Comment by TravisEz13** [comment]
travis-ci finished and did not update:  https://travis-ci.org/PowerShell/PowerShell/builds/426798974?utm_source=github_status&utm_medium=notification

**Pull request merged**